core/src/main/scala/org/apache/spark/scheduler/TaskResult.scala

```diff
@@ -68,6 +73,17 @@ private[spark] class DirectTaskResult[T](
       }
       accumUpdates = _accumUpdates
     }
+
+    val numMetrics = in.readInt
+    if (numMetrics == 0) {
+      metricPeaks = Array.empty
+    } else {
+      val _metricPeaks = new ArrayBuffer[Long]
+      for (i <- 0 until numMetrics) {
+        _metricPeaks += in.readLong
+      }
+      metricPeaks = _metricPeaks.toArray
```

Comment:
since you know the size you are reading, and you're going to create an array in the end, you should just create the array in the first place.

```
val metricPeaks = new Array[Long](numMetrics)
(0 until numMetrics).foreach { idx =>
  metricPeaks(idx) = in.readLong
}
```